# Marketing Campaign Response Modelling

## Literature Review

Author:
Minal Sadiq

Dataset:
UCI Bank Marketing Dataset (Moro et al., 2014)

Objective:
Predict customers who are most likely to subscribe to a marketing campaign in order to improve campaign targeting and reduce marketing costs.

In [1]:
import pandas as pd

df = pd.read_csv("../data/raw/bank-additional-full.csv", sep=";")

df.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


Dataset Information

In [2]:
df.shape

(41188, 21)

Columns

In [3]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

Data Types

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

Missing Values

In [5]:
df.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

Duplicates

In [6]:
df.duplicated().sum()

np.int64(12)

Target Distribution

In [7]:
df["y"].value_counts()

df["y"].value_counts(normalize=True)*100

y
no     88.734583
yes    11.265417
Name: proportion, dtype: float64

Statistical Summary

In [8]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,41188.0,40.024060,10.421250,17.000,32.000,38.000,47.000,98.000
duration,41188.0,258.285010,259.279249,0.000,102.000,180.000,319.000,4918.000
campaign,41188.0,2.567593,2.770014,1.000,1.000,2.000,3.000,56.000
pdays,41188.0,962.475454,186.910907,0.000,999.000,999.000,999.000,999.000
previous,41188.0,0.172963,0.494901,0.000,0.000,0.000,0.000,7.000
emp.var.rate,41188.0,0.081886,1.570960,-3.400,-1.800,1.100,1.400,1.400
cons.price.idx,41188.0,93.575664,0.578840,92.201,93.075,93.749,93.994,94.767
cons.conf.idx,41188.0,-40.502600,4.628198,-50.800,-42.700,-41.800,-36.400,-26.900
euribor3m,41188.0,3.621291,1.734447,0.634,1.344,4.857,4.961,5.045
nr.employed,41188.0,5167.035911,72.251528,4963.600,5099.100,5191.000,5228.100,5228.100


Categorical Summary

In [9]:
df.describe(include="object").T

,count,unique,top,freq
job,41188,12,admin.,10422
marital,41188,4,married,24928
education,41188,8,university.degree,12168
default,41188,3,no,32588
housing,41188,3,yes,21576
loan,41188,3,no,33950
contact,41188,2,cellular,26144
month,41188,10,may,13769
day_of_week,41188,5,thu,8623
poutcome,41188,3,nonexistent,35563


Generate Professional Profile Report

In [10]:
from ydata_profiling import ProfileReport

profile = ProfileReport(
    df,
    title="Marketing Campaign Profiling Report",
    explorative=True
)

profile.to_file("../Reports/Profiling/marketing_profile.html")

d:\Important Files\Marketing Campaign Response Modelling\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
C:\Users\User\AppData\Local\Temp\ipykernel_12964\3401690907.py:1: DeprecationWarning: 
    `import ydata_profiling` is deprecated and will not receive more updates. 
    Please install fg-data-profiling via `pip install fg-data-profiling` and use `import data_profiling` instead.
    
  from ydata_profiling import ProfileReport
Export report to file: 100%|██████████| 1/1 [00:00<00:00, 42.59it/s]


# Business Problem Statement

Marketing campaigns are expensive, and contacting every customer is inefficient.

The objective of this project is to develop a machine learning model that predicts whether a customer will subscribe to a term deposit after a direct marketing campaign.

The model will help financial institutions:

- Improve campaign targeting
- Reduce marketing costs
- Increase conversion rates
- Maximize Return on Investment (ROI)
- Support data-driven decision making

Dataset Information

In [11]:
print("Dataset Overview")

print(f"Rows      : {df.shape[0]:,}")
print(f"Columns   : {df.shape[1]}")
print(f"Features  : {df.shape[1]-1}")
print(f"Target    : y")

Dataset Overview
Rows      : 41,188
Columns   : 21
Features  : 20
Target    : y


Numerical vs Categorical Features

In [12]:
categorical_cols = df.select_dtypes(include="object").columns.tolist()
numerical_cols = df.select_dtypes(exclude="object").columns.tolist()

print("Categorical Features:", len(categorical_cols))
print(categorical_cols)

print()

print("Numerical Features:", len(numerical_cols))
print(numerical_cols)

Categorical Features: 11
['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome', 'y']

Numerical Features: 10
['age', 'duration', 'campaign', 'pdays', 'previous', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed']


Target Variable Summary

In [13]:
target_summary = pd.DataFrame({
    "Count": df["y"].value_counts(),
    "Percentage": (df["y"].value_counts(normalize=True) * 100).round(2)
})

target_summary

,Count,Percentage
y,,
no,36548,88.73
yes,4640,11.27


# Initial Observations

### Dataset Characteristics

- 41,188 customer records
- Binary classification problem
- Target variable: y (Yes / No)
- Mix of numerical and categorical variables
- No missing values detected
- Dataset contains "unknown" categories instead of NaN values
- Class distribution is imbalanced
- Marketing campaign data collected by a Portuguese bank

Save Basic Dataset Information

In [14]:
dataset_info = pd.DataFrame({
    "Feature": df.columns,
    "Data Type": df.dtypes.astype(str).values
})

dataset_info.to_csv("../reports/dataset_information.csv", index=False)

print("Dataset information saved successfully.")

Dataset information saved successfully.


Save Target Summary

In [15]:
target_summary.to_csv("../reports/target_summary.csv")

print("Target summary saved successfully.")

Target summary saved successfully.


# Conclusion

The dataset has been successfully loaded and inspected.

Key findings:

- Dataset contains 41,188 observations.
- No missing values are present.
- Several categorical variables contain "unknown" values.
- The target variable is imbalanced.
- The dataset is suitable for marketing response prediction.

The next notebook will focus on:

- Exploratory Data Analysis (EDA)
- Class imbalance analysis
- Feature distributions
- Correlation analysis
- Data cleaning